File: 3957-maximum-sum-of-m-non-overlapping-subarrays-ii/3957-maximum-sum-of-m-non-overlapping-subarrays-ii.java
class Solution {
    long[] pref;
    int n;
    long[] dp;
    int[] cnt;
    int[] dq; 
    
    class Result {
        long val;
        int cnt;
        Result(long val, int cnt) {
            this.val = val;
            this.cnt = cnt;
        }
    }
    
    public long maximumSum(int[] nums, int m, int l, int r) {
        n = nums.length;
        pref = new long[n + 1];
        for (int i = 0; i < n; i++) {
            pref[i + 1] = pref[i] + nums[i];
        }
        
        dp = new long[n + 1];
        cnt = new int[n + 1];
        dq = new int[n + 1];
        
        long maxSingle = (long) -1e18;
        int head = 0, tail = 0;
        for (int i = l; i <= n; i++) {
            int p = i - l;
            while (tail > head && pref[dq[tail - 1]] >= pref[p]) {
                tail--;
            }
            dq[tail++] = p;
            while (tail > head && dq[head] < i - r) {
                head++;
            }
            maxSingle = Math.max(maxSingle, pref[i] - pref[dq[head]]);
        }
        
        if (maxSingle <= 0) {
            return maxSingle;
        }
        
        Result res0 = check(0, l, r);
        if (res0.cnt <= m) {
            return res0.val;
        }
        
        long low = 0, high = (long) 2e14, ansC = 0;
        while (low <= high) {
            long mid = low + (high - low) / 2;
            if (check(mid, l, r).cnt >= m) {
                ansC = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        return check(ansC, l, r).val + m * ansC;
    }
    
    private Result check(long C, int l, int r) {
        for (int i = 0; i <= n; i++) {
            dp[i] = 0;
            cnt[i] = 0;
        }
        int head = 0, tail = 0;
        
        for (int i = l; i <= n; i++) {
            int p = i - l;
            
            while (tail > head) {
                int back = dq[tail - 1];
                long val1 = dp[p] - pref[p];
                long val2 = dp[back] - pref[back];
                if (val1 > val2 || (val1 == val2 && cnt[p] > cnt[back])) {
                    tail--;
                } else {
                    break;
                }
            }
            dq[tail++] = p;
            
            while (tail > head && dq[head] < i - r) {
                head++;
            }
            
            dp[i] = dp[i - 1];
            cnt[i] = cnt[i - 1];
            
            if (tail > head) {
                int best = dq[head];
                long takeVal = dp[best] - pref[best] + pref[i] - C;
                int takeCnt = cnt[best] + 1;
                
                if (takeVal > dp[i] || (takeVal == dp[i] && takeCnt > cnt[i])) {
                    dp[i] = takeVal;
                    cnt[i] = takeCnt;
                }
            }
        }
        return new Result(dp[n], cnt[n]);
    }
}

File: 3957-maximum-sum-of-m-non-overlapping-subarrays-ii/3957-maximum-sum-of-m-non-overlapping-subarrays-ii.cpp
class Solution {
public:
    struct State {
        long long val;
        int cnt;
    };

    bool better(State a, State b, bool preferMore) {
        if (a.val != b.val) return a.val > b.val;
        return preferMore ? a.cnt > b.cnt : a.cnt < b.cnt;
    }

    long long maximumSum(vector<int>& nums, int m, int l, int r) {
        int n = nums.size();

        vector<long long> pref(n + 1, 0);
        long long bound = 1;
        for (int i = 0; i < n; i++) {
            pref[i + 1] = pref[i] + nums[i];
            bound += llabs((long long)nums[i]);
        }

        auto bestOne = [&]() {
            deque<int> dq;
            long long best = LLONG_MIN / 4;

            for (int i = 1; i <= n; i++) {
                int add = i - l;
                if (add >= 0) {
                    while (!dq.empty() && pref[dq.back()] >= pref[add]) dq.pop_back();
                    dq.push_back(add);
                }

                while (!dq.empty() && dq.front() < i - r) dq.pop_front();

                if (!dq.empty()) {
                    best = max(best, pref[i] - pref[dq.front()]);
                }
            }

            return best;
        };

        long long one = bestOne();
        if (one <= 0) return one;

        auto calc = [&](long long cost, bool preferMore) {
            vector<State> dp(n + 1, {0, 0});
            deque<int> dq;

            auto key = [&](int idx) {
                return State{dp[idx].val - pref[idx], dp[idx].cnt};
            };

            for (int i = 1; i <= n; i++) {
                int add = i - l;
                if (add >= 0) {
                    while (!dq.empty() && better(key(add), key(dq.back()), preferMore)) {
                        dq.pop_back();
                    }
                    dq.push_back(add);
                }

                while (!dq.empty() && dq.front() < i - r) dq.pop_front();

                dp[i] = dp[i - 1];

                if (!dq.empty()) {
                    State bestStart = key(dq.front());
                    State cand = {
                        pref[i] - cost + bestStart.val,
                        bestStart.cnt + 1
                    };

                    if (better(cand, dp[i], preferMore)) {
                        dp[i] = cand;
                    }
                }
            }

            return dp[n];
        };

        State freeTake = calc(0, false);
        if (freeTake.cnt <= m) return freeTake.val;

        long long lo = 0, hi = bound;
        while (lo < hi) {
            long long mid = (lo + hi + 1) / 2;
            if (calc(mid, true).cnt >= m) {
                lo = mid;
            } else {
                hi = mid - 1;
            }
        }

        State res = calc(lo, true);
        return res.val + lo * m;
    }
};